lib/protobuf/verifier.ex

```diff
@@ -0,0 +1,265 @@
+    end
+  end
+
+  defp verify_fields(fields, syntax, struct, oneofs) do
+    Enum.map(fields, fn %{name_atom: name, oneof: oneof} = prop ->
+      val =
+        if oneof do
+          oneofs[name]
+        else
+          case struct do
+            %{^name => v} -> v
+            _ -> nil
+          end
+        end
+
+      if skip_field?(syntax, val, prop) || skip_enum?(prop, val) do
+        :ok
+      else
+        verify_field(class_field(prop), val, prop)
+        |> wrap_error(struct, prop)
+      end
+    end)
+    |> first_non_ok_value_if_present()
+  end
+
+  defp wrap_error(:ok, _struct, _prop), do: :ok
+
+  defp wrap_error({:error, msg}, struct, prop) do
+    wrapped_msg =
+      "Got error when verifying the values of #{inspect(struct.__struct__)}##{prop.name_atom}: #{
+        msg
+      }"
+
+    {:error, wrapped_msg}
+  end
+
+  def skip_field?(_syntax, val, %{type: type, options: options} = prop) when not is_nil(options),
+    do: FieldOptionsProcessor.skip_verify?(type, val, prop, options)
+
+  def skip_field?(_, [], _), do: true
+  def skip_field?(_, v, _) when map_size(v) == 0, do: true
+  def skip_field?(:proto2, nil, %{optional?: true}), do: true
+  def skip_field?(:proto3, nil, _), do: true
+  def skip_field?(_, _, _), do: false
+
+  @spec verify_field(atom, any, FieldProps.t()) :: :ok | {:error, String.t()}
+  defp verify_field(
+         :normal,
+         val,
+         %{type: type, repeated?: is_repeated} = prop
+       ) do
+    repeated_or_not(val, is_repeated, fn v ->
+      if is_nil(prop.options) do
+        verify_type(type, v)
+      else
+        FieldOptionsProcessor.verify_type(type, v, prop.options)
+      end
+    end)
+    |> first_non_ok_value_if_present()
+  end
+
+  # The guard ensures that val's type matches the is_repeated or is_map parameters
+  defp verify_field(
+         :embedded,
+         val,
+         %{repeated?: is_repeated, map?: is_map, type: type} = prop
+       )
+       when (is_repeated and is_list(val)) or (is_map and is_map(val)) or
+              (not is_repeated and not is_map) do
+    repeated = is_repeated or is_map
+
+    repeated_or_not(val, repeated, fn v ->
+      v = if is_map, do: struct(prop.type, %{key: elem(v, 0), value: elem(v, 1)}), else: v
+
+      if is_nil(prop.options) do
+        verify(type, v)
+      else
+        FieldOptionsProcessor.verify_type(type, v, prop.options)
+      end
+    end)
+    |> first_non_ok_value_if_present()
+  end
+
+  # A catchall for params that don't match the verify_field(:embedded) guard above
+  defp verify_field(:embedded, val, prop),
+    do:
+      {:error,
+       "Got a value: #{inspect(val)} that isn't a map or list for the repeated or map field #{
+         prop.name_atom
+       }"}
+
+  defp repeated_or_not(val, true = _repeated, func), do: Enum.map(val, func)
+  defp repeated_or_not(val, false = _repeated, func), do: [func.(val)]
+
+  # NOTE: This can aggregate all of the errors if we want
```

Comment:
I think it might be nice to show all errors at once. Thats what other validators like optimal https://github.com/albert-io/optimal do. Wydt?

Reply:
I think that'd be handy, let me know what you think of the new interface